gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoRestSerializer.java

```diff
@@ -0,0 +1,105 @@
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.httpjson;
+
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.Message;
+import com.google.protobuf.util.JsonFormat;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.util.List;
+import java.util.Map;
+
+public class ProtoRestSerializer<RequestT extends Message> {
+  private ProtoRestSerializer() {}
+
+  public static <RequestT extends Message> ProtoRestSerializer<RequestT> create() {
+    return new ProtoRestSerializer<>();
+  }
+
+  public String toJson(RequestT message) {
+    try {
+      return JsonFormat.printer().print(message);
+    } catch (InvalidProtocolBufferException e) {
+      throw new RuntimeException("Failed to serialize message to JSON", e);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  public RequestT fromJson(InputStream message, Message.Builder builder) {
+    try (Reader json = new InputStreamReader(message)) {
+      JsonFormat.parser().ignoringUnknownFields().merge(json, builder);
+      return (RequestT) builder.build();
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to parse response message", e);
+    }
+  }
+
+  public void putPathParam(Map<String, String> fields, String fieldName, Object fieldValue) {
+    if (isDefaultValue(fieldName, fieldValue)) {
+      return;
+    }
+    fields.put(fieldName, String.valueOf(fieldValue));
+  }
+
+  public void putQueryParam(Map<String, List<String>> fields, String fieldName, Object fieldValue) {
+    if (isDefaultValue(fieldName, fieldValue)) {
```

Comment:
(but I realize I'm not clear on where these `put*` methods are invoked)

Reply:
I believe I added this because it was literally breaking the execution of the generated clients. I guess it is especially relevant for strings, which have empty strings as default values, and this leads do path params being something like `?path_param=&path_param2=`.

We will get back to default parameters either way. To match default values doc, plust we will have to revisit it if/when we move to using field presence feature of proto3. Please lets keep this as is for now, because it was added to fix specific failures and I have the whole toolchain tested with this present.

Reply:
That's great context, could you please add a comment?

Reply:
Sounds reasonable, but let's flag this more prominently. Could you add below your existing comment something like  `TODO: Revisit this approach to ensure proper default-value handling as per design.`